[`ssort`] Implement statement sorting

JackAshwell11 opened the pull request:
## Summary

<!-- What's the purpose of the change? What does it do, and why? -->

Closes #3946 and #2436

- Adds a new SS001 (unsorted statements) rule allowing a file to be automatically sorted based on the dependencies between its functions/methods/classes. This also comes with an extra linter option `narrative_order` allowing the statement ordering to be reversed meaning the calling functions are defined before the dependency functions.

Note that this PR intentionally avoids sorting statements inside classes such as the docstring/special attribute/lifecycle methods/etc which ssort implements (https://github.com/bwhmather/ssort?tab=readme-ov-file#output) as these fall under #2425 which needs further discussion.

## Test Plan

<!-- How was it tested? -->

- Multiple Python files have been added to test the diagnostic reporting of the two rules in both dependency and narrative ordering mode.
- Additional Rust unit tests have been added to test extra functionality not directly covered by the Python tests.

I've checked the `ruff-ecosystem` results and each change looks to be correct with this new rule (although there will be many rule violations).

JackAshwell11 committed:
Initial commit of setting up ssort rules and linter compatability

JackAshwell11 committed:
Introduced a dependency graph and a topological sort for analysing and resolving a function's dependencies with basic autofix support

JackAshwell11 committed:
Refactored and cleaned up ssort's internals, so every statement now becomes a `Node` allowing top-level statements, which aren't functions or classes, to impact the dependency ordering

This also makes the replacement string construction much easier as we can emit each node in order as it is fully sorted at that point

JackAshwell11 committed:
Improved ssort replacement string generation to preserve original formatting and use the locator API

JackAshwell11 committed:
Added support for sorting inside classes by recursively calling `organize_suite()` during string replacement

JackAshwell11 committed:
Added support for narrative ordering in ssort controllable via the linter settings

JackAshwell11 committed:
Added a new rule for detecting cycles in function/method calls

JackAshwell11 committed:
Added Python test file fixtures to test the correctness of the ssort rules

Added comprehensive Rust tests for functionality not directly covered by the Python tests

ntBre commented on crates/ruff_linter/resources/test/fixtures/ssort/async_function.py:
Wow that's a lot of new test files! Thanks for being thorough here, but is there any chance we could condense this down a bit? I haven't taken a look at the implementation yet, but something we often do is create outer functions or classes to contain test cases. For example, could we test the same sorting behavior if this file were like:

```py
class C:
	def function():
	    return dependency()
	
	
	def dependency():
	    return True
	
	
	async def async_function():
	    return function()
```

and then another file could be:

```py
class D:
    # another test case
    ...
```

or do they really need to be separate files?

ntBre reviewed (commented):
Thanks! I need to do a much deeper review at some point, just a couple of passing thoughts while scrolling through to approve the workflows.

JackAshwell11 committed:
Removed `FunctionCallCycle` as it may be controversial

MichaReiser commented on crates/ruff_linter/src/rules/ssort/dependencies.rs:
```diff
@@ -0,0 +1,785 @@
+                    if value.id.as_str() == "self" {
+                        self.names.insert(&attr.attr.id);
+                    }
+                }
```
Don't we end up visiting `attr.value` twice? Or did you want to return early when seeing an attribute expression?

JackAshwell11 replied:
This is unintentional, I'll fix it. Thanks 👍

MichaReiser commented on crates/ruff_linter/src/rules/ssort/dependencies.rs:
```diff
@@ -0,0 +1,785 @@
+    ///
+    /// ## Arguments
+    /// * `expr` - The expression node to visit.
+    fn visit_expr(&mut self, expr: &'a Expr) {
```
Does this visitor correctly handle scopes? What if two functions define the same variables but they're nested within each other? What about type vars?

JackAshwell11 replied:
From my understanding, `ssort` is intentionally scope-agnostic as it only tracks text references, so as long as the text reference matches another function, this should work.

If two functions define the same variables but they're nested then this shouldn't handle they would be considered a separate suite (although I'd argue that is a problem with your code as you shouldn't name two things with the same name).

I'm not sure if type variables will work as I don't think `visit_expr()` handles annotations, but I could be wrong

bwhmather replied:
> From my understanding, ssort is intentionally scope-agnostic as it only tracks text references, so as long as the text reference matches another function, this should work.

Hello - ssort author here.  ssort is not scope-agnostic but we sort of do it backwards.  In the nested function case, the shared variable will be included in the "requirements" (unresolved "dependencies") for the inner function but will be masked out of the requirements for the outer one.  Relevant bit probably [here](https://github.com/bwhmather/ssort/blob/cc3758286bb788c97c6619e99f52d49cfb292893/src/ssort/_requirements.py#L88-L100).

bwhmather replied:
> How does ssort ensure that type annotations don't break 

Not sure if it answers your question, but in ssort we traverse annotations normally when looking for "requirements" and include any bindings that they introduce in the function scope when pruning the function's requirements.

bwhmather replied:
Btw, ruff and ssort are both MIT licenced.  You're more than welcome to copy as much as you like.  In particular the test suite has cases to cover _a lot_ of weird edge cases like this.

JackAshwell11 commented on crates/ruff_linter/src/checkers/statements.rs:
Should be fixed in latest

JackAshwell11 committed:
Added code review fixes:
- Optimised `DependencyVisitor::visit_expr()` so it doesn't slow Ruff down as much.
- Removed some problematic Python test files.
- Fixed a bug where `attr.value` is visited twice.
- Fixed a bug where type annotations would cause sorting to break.
- Fixed all clippy warnings.

ntBre commented:
Thank you again for all of your work here! It's going to take me some time to digest this and review, but I have it on my todo list for the month.

Happy new year!

ntBre committed:
Merge branch 'main' into jack/ssort

ntBre committed:
update schema and linter sort order

ntBre commented on crates/ruff_linter/src/codes.rs:
```diff
@@ -1194,6 +1192,9 @@ pub fn code_to_rule(linter: Linter, code: &str) -> Option<(RuleGroup, Rule)> {
         (Flake8Logging, "015") => rules::flake8_logging::rules::RootLoggerCall,
 
+        // ssort
+        (SSort, "001") => rules::ssort::rules::UnsortedStatements,
```
I'd be curious to get other thoughts on this, but I'd probably just make this a `RUF` rule. We're hoping to move away from having all of these separate linters before long anyway (#1774), and I don't think it's worth adding a new linter prefix for one rule.

ntBre commented on crates/ruff_linter/src/rules/ssort/rules/organize_statements.rs:
```diff
@@ -0,0 +1,609 @@
+///
+/// ## Returns
+/// A string representing the replacement text for the statement.
+fn get_replacement_text(locator: &Locator, stmt: &Stmt, narrative_order: bool) -> String {
```
We might want to use a `Cow` here to avoid allocating a `String` in cases where we're just returning a reference into the existing source:


```suggestion
fn get_replacement_text<'a>(locator: &'a Locator, stmt: &Stmt, narrative_order: bool) -> Cow<'a, str> {
```

ntBre commented on crates/ruff_linter/src/rules/ssort/dependencies.rs:
```diff
@@ -0,0 +1,1222 @@
+///
+/// ## Returns
+/// A vector of nodes representing the statements in the suite.
+pub(super) fn nodes_from_suite(suite: &'_ [Stmt]) -> Vec<Node<'_>> {
```
```suggestion
pub(super) fn nodes_from_suite(suite: &[Stmt]) -> Vec<Node> {
```

Not sure if it will let you elide the second one, but the first one should definitely be okay.

ntBre commented on crates/ruff_linter/src/rules/ssort/dependencies.rs:
```diff
@@ -0,0 +1,1222 @@
+                index,
+                name,
+                stmt,
+                movable: name.is_some(),
```
Do we need to track this field separately if it's just `name.is_some()`?

ntBre committed:
delete unreferenced snapshots

ntBre reviewed (commented):
Thanks for your work here and for your patience! I still need to wrap my head fully around the graph algorithm, but I did a deeper dive today and came up with some suggestions for simplification. In particular, I think if we can narrow down the (very thorough!) tests and restructure them so that the important features are more evident, it will be a bit easier to review the logical aspects of the PR.

JackAshwell11 commented on crates/ruff_linter/src/rules/ssort/rules/organize_statements.rs:
```diff
@@ -0,0 +1,609 @@
+    };
+
+    // Construct the replacement text by sorting the class body
+    let inner_replacement = organize_suite(&class_def.body, locator, narrative_order);
```
The rule should recurse into classes and sort them as well. It shouldn't recurse into a function though as we only want to move around class/function blocks (and anything else that can be moved so function statements can't be moved)

JackAshwell11 commented on crates/ruff_linter/src/rules/ssort/rules/organize_statements.rs:
```diff
@@ -0,0 +1,609 @@
+# Comment 2
+
+def bar():
+    pass
```
What do you mean here?

JackAshwell11 commented on crates/ruff_linter/src/rules/ssort/dependencies.rs:
```diff
@@ -0,0 +1,1222 @@
+                    .keywords
+                    .iter()
+                    .for_each(|kw| self.visit_expr(&kw.value));
+            }
```
Its been a while since I've looked at this, but I think @MichaReiser had some suggestions which required some custom functionality here?

JackAshwell11 committed:
Added code review fixes

bwhmather commented on crates/ruff_linter/src/rules/ssort/rules/organize_statements.rs:
```diff
@@ -0,0 +1,590 @@
+        // Append the separator unless this is the last sorted node
+        if i < sorted.len() - 1 {
+            replacement.push_str(separator_for(suite, locator, node_idx));
+        }
```
I might be misreading this but in ssort we usually attach comments and whitespace to the following statement, not the preceding statement.

I _think_ this will format:
```python
# Documentation for a.
a = lambda : b
# Documentation for b.
b = 1
```
as:
```python
# Documentation for a.
b = 1
a = lambda : b
# Documentation for b.
```
